lib/MongoDB/ClientSession.pm

```diff
@@ -215,6 +259,242 @@ sub advance_operation_time {
+
+=cut
+
+sub commit_transaction {
+    my $self = shift;
+
+    MongoDB::TransactionError->throw("No transaction started")
+        if $self->_transaction_state eq 'none';
+
+    # Error message tweaked to use our function names
+    MongoDB::TransactionError->throw("Cannot call commit_transaction after calling abort_transaction")
+        if $self->_transaction_state eq 'aborted';
+
+    # Commit can be called multiple times - even if the transaction completes
+    # correctly. Setting this here makes sure we dont increment transaction id
+    # until after another command has been called using this session
+    $self->_set__active_transaction( 1 );
+
+    eval {
+        $self->_send_end_transaction_command( 'committed', [ commitTransaction => 1 ] );
+    };
+    if ( my $err = $@ ) {
+        # catch and re-throw after retryable errors
+        # TODO maybe need better checking logic that theres actually an error code in output?
+        my $err_code;
+        if ( $err->can('result') ) {
+            if ( $err->result->can('output') ) {
+                $err_code = $err->result->output->{codeName};
+                $err_code ||= $err->result->output->{writeConcernError}
+                  ? $err->result->output->{writeConcernError}->{codeName}
+                  : ''; # Empty string just in case
+            }
+        }
+        # If its a write concern error, retrying a commit would still error
+        unless ( defined( $err_code ) && grep { $_ eq $err_code } qw/
+            CannotSatisfyWriteConcern
+            UnsatisfiableWriteConcern
+            UnknownReplWriteConcern
+            NoSuchTransaction
+        / ) {
+            push @{ $err->error_labels }, 'UnknownTransactionCommitResult';
+        }
+        die $err;
+    }
+
+    return;
+}
+
+=method abort_transaction
+
+Abort the current transaction. This will use the writeConcern set on this transaction.
+
+=cut
+
+sub abort_transaction {
+    my $self = shift;
+
+    MongoDB::TransactionError->throw("No transaction started")
+        if $self->_in_transaction_state( 'none' );
+
+    # Error message tweaked to use our function names
+    MongoDB::TransactionError->throw("Cannot call abort_transaction after calling commit_transaction")
+        if $self->_in_transaction_state( 'committed' );
+
+    # Error message tweaked to use our function names 
+    MongoDB::TransactionError->throw("Cannot call abort_transaction twice")
+        if $self->_in_transaction_state( 'aborted' );
+
+    $self->_send_end_transaction_command( 'aborted', [ abortTransaction => 1 ] );
+
+    return;
+}
+
+sub _send_end_transaction_command {
+    my ( $self, $end_state, $command ) = @_;
+
+    # Only need to send commit command if the transaction actually sent anything
+    if ( $self->_has_transaction_operations ) {
+
+        # Must set state before running the op as otherwise it wont be retried
+        $self->_set__transaction_state( $end_state );
```

Comment:
This state is supposed to change regardless of whether there are operations.

Reply:
It's also down below at 428.  Maybe it should be done once before the loop, as the state changes whether errors occur or not.